File: devops-scm-spring-boot-starter/src/main/java/com/tencent/devops/scm/spring/manager/ScmCommandManager.java
package com.tencent.devops.scm.spring.manager;

import com.tencent.devops.scm.api.ScmCommand;
import com.tencent.devops.scm.api.pojo.repository.ScmProviderRepository;
import com.tencent.devops.scm.spring.properties.ScmProviderProperties;
import java.util.List;

public class ScmCommandManager {

    private final List<ScmCommandFactory> commandFactories;

    public ScmCommandManager(List<ScmCommandFactory> commandFactories) {
        this.commandFactories = commandFactories;
    }

    public ScmCommand build(ScmProviderProperties properties) {
        for (ScmCommandFactory factory : commandFactories) {
            if (factory.support(properties)) {
                return factory.build(properties);
            }
        }
        throw new NoSuchScmProviderException(properties.getProviderCode());
    }

    public void remoteInfo(ScmProviderProperties properties, ScmProviderRepository repository) {
        build(properties).remoteInfo(repository);
    }
}

File: devops-scm-spring-boot-starter/src/main/kotlin/com/tencent/devops/scm/spring/manager/ScmCommandManager.kt
package com.tencent.devops.scm.spring.manager

import com.tencent.devops.scm.api.ScmCommand
import com.tencent.devops.scm.api.pojo.repository.ScmProviderRepository
import com.tencent.devops.scm.spring.properties.ScmProviderProperties

class ScmCommandManager(
    private val commandFactories: List<ScmCommandFactory>
) {

    fun build(properties: ScmProviderProperties): ScmCommand {
        return commandFactories.firstOrNull { it.support(properties) }
            ?.build(properties)
            ?: throw NoSuchScmProviderException(properties.providerCode!!)
    }

    fun remoteInfo(properties: ScmProviderProperties, repository: ScmProviderRepository) {
        build(properties).remoteInfo(repository)
    }
}
